### PR TITLE
OSSM-2446 Introduce profiles

### PR DESCRIPTION
**Pull request by luksa**

### CODE REVIEW

File: hack/download-charts.sh
Comment (jwendell) [inline hack/download-charts.sh]: YQ is already present is the istio upstream builder image. We should switch to it ASAP. @bmangoen is working on this.

File: hack/download-charts.sh
Comment (jwendell) [inline hack/download-charts.sh:53]: Do we really need to do that? I'm worried this operator becomes 2.x with lots of download/patch/sed/hack boilerplate code.

Can't we just define our own set of profiles, based on upstream?

File: api/v1alpha1/istio_types.go
Comment (jwendell) [inline api/v1alpha1/istio_types.go:37]: What about the openshift profile? Keep in mind that this is an upstream project, we can't default to openshift. Or at least we shouldn't, IMO.

File: hack/download-charts.sh
Comment (luksa) [reply hack/download-charts.sh:53]: We do. We need to convert from `IstioOperator.install.istio.io` to `Istio.operator.istio.io` and remove the fields that we don't support (yet?), such as `components`, `meshConfig`, `hub`, and `tag`.

That's exactly what we're doing. We're defining our own set based on upstream.

File: api/v1alpha1/istio_types.go
Comment (luksa) [reply api/v1alpha1/istio_types.go:37]: We don't default to OpenShift, but we do override some values [here](https://github.com/Maistra/istio-operator/blob/fff1734ab6053c8a43b8624b386108d19dc3a14e/pkg/strategy/maistra_30_strategy.go#L20-L47). The plan is to eventually remove those overrides.

Comment (jwendell) [reply api/v1alpha1/istio_types.go:37]: this is wrong, no? We shouldn't be overriding values such as cni.enabled I think.

File: hack/download-charts.sh
Comment (jwendell) [reply hack/download-charts.sh:53]: Sorry, I meant, should we just have static files in our repo instead of patching istio/istio files?

File: api/v1alpha1/istio_types.go
Comment (luksa) [reply api/v1alpha1/istio_types.go:37]: I agree. But this is how it was done initially and changing it isn't part of this PR.

File: hack/download-charts.sh
Comment (luksa) [reply hack/download-charts.sh:53]: Why? Then we'd have to update them manually whenever they change upstream. IMHO we should support upstream profiles, not create our own.

File: api/v1alpha1/istio_types.go
Comment (jwendell) [reply api/v1alpha1/istio_types.go:37]: I got confused on why isn't openshift listed here. Plus, do we want to hardcode these values in the doc or just say that we accept any value present in istio/istio/manifests/profiles?

File: hack/download-charts.sh
Comment (jwendell) [reply hack/download-charts.sh:53]: We are already getting only the spec.values map within `getValues()`. Can't we just use the raw upstream file as-is and ignore everything that's not `spec.values` in it?

Comment (luksa) [reply hack/download-charts.sh:53]: Using the upstream file would be misleading and confusing because:
- the upstream file is an IstioOperator manifest
- ignoring fields will make things confusing and misleading; users will waste time figuring out why something isn't working

Also, this probably isn't the final state of these files. But we can't yet know what the final state will be. The way these files are transformed in this PR is what makes sense currently. We will most likely change them eventually.

File: hack/download-charts.sh
Comment (luksa) [reply hack/download-charts.sh]: I'll put this PR on hold until @bmangoen's PR is merged.

File: Makefile
Comment (jwendell) [inline Makefile]: With the changes in #1352 we don't need this. Any `make` commands will run within the build container by default, which already contains yq.

File: controllers/istio_controller.go
Comment (jwendell) [inline controllers/istio_controller.go:341]: Is it worth to make this fallback value configurable somehow? In community builds, it defaults to `default` but product builds would default to `openshift`?

File: controllers/istio_controller.go
Comment (jwendell) [inline controllers/istio_controller.go:377]: Does it play well with the `empty` profile?

File: controllers/istio_controller_test.go
Comment (jwendell) [inline controllers/istio_controller_test.go:305]: perhaps out of the scope for this pr, but we should avoid hard coding `v3.0` everywhere

File: resources/v3.0/profiles/minimal.yaml
Comment (jwendell) [inline resources/v3.0/profiles/minimal.yaml:2]: Because we got rid of everything !spec.values, this profile is identical to empty. Do we want to keep both? Does it make sense to have them at all?

**Review by jwendell — COMMENTED** [review]
Do you think it's worth to add integration tests for each of the profiles? To make sure they are valid?

File: Makefile
Comment (luksa) [reply Makefile]: removed

File: controllers/istio_controller.go
Comment (luksa) [reply controllers/istio_controller.go:341]: let's think about this when we clean up the openshift-specific stuff in the maistra30strategy.go file.

File: controllers/istio_controller.go
Comment (luksa) [reply controllers/istio_controller.go:377]: It should. For the empty profile, this returns a nil map. This nil map is then only used in `mergeValues()`, which can safely handle the nil value here: https://github.com/Maistra/istio-operator/blob/0d82b6ff8fca6e874e01c726efef03f791dc7124/controllers/istio_controller.go#L391

File: controllers/istio_controller_test.go
Comment (luksa) [reply controllers/istio_controller_test.go:305]: This "v3.0" instance is not a problem, since it's only used within the TestApplyProfile function. It could be any string, as long as the same string is used in the testcases within TestApplyProfile.

File: resources/v3.0/profiles/minimal.yaml
Comment (luksa) [reply resources/v3.0/profiles/minimal.yaml:2]: Let's keep them for now, since we'll most likely introduce the `components` field, at which point these two profiles will no longer be identical.

**Review by jwendell — APPROVED** [review]

**Comment by jwendell** [comment]
We can think of adding integration tests for those profiles later

**Pull request merged**
